### ISSUE TITLE
Is CPU jsrun line correct?

### ISSUE THREAD

**Issue by oksanaguba** [E3SM-Project/scream#1723]
Running from today's master
```
commit fcf28ecbac2c4d5aab7c727b4834f89433b20ae2 (HEAD -> master, origin/master, origin/HEAD)
Merge: 587d6c3fc5 679a5812b8
Author: James Foucar <jgfouca@sandia.gov>
Date:   Mon Jun 6 10:16:09 2022 -0600

    Merge pull request #1687 from E3SM-Project/xyuan/scream_summit_cpu
```
preview run looks like this (jsrun: look at **--bind packed:smt:2** )
```
[onguba@login2.summit v1-cpu-ne120-nnodes40-june06-check-master]$ ./preview_run 
CASE INFO:
  nodes: 40
  total tasks: 1680
  tasks per node: 42
  thread count: 1
  ngpus per node: 0

BATCH INFO:
  FOR JOB: case.run
    ENV:
      Setting Environment CPU_PER_RS=21
      Setting Environment ESSL_PATH=/sw/summit/essl/6.3.0/essl/6.3
      Setting Environment GPU_PER_RS=0
      Setting Environment HDF5_PATH=/sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-7.5.0/hdf5-1.10.7-ycmyphhsuprml7oh7bxgcb2np3v732cl
      Setting Environment JSRUN_THREAD_VARS= 
      Setting Environment LTC_PRT=cpu-cpu
      Setting Environment NETCDF_C_PATH=/sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-7.5.0/netcdf-c-4.8.0-pwi4jbrnwv4lrrjxdu5czbos5uvvjgvr
      Setting Environment NETCDF_FORTRAN_PATH=/sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-7.5.0/netcdf-fortran-4.4.5-p6rxtzavwducqg3mihef2xekcywgh3yk
      Setting Environment NUM_RS=80
      Setting Environment OMP_NUM_THREADS=1
      Setting Environment PGI_ACC_POOL_ALLOC=0
      Setting Environment PNETCDF_PATH=/sw/summit/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-7.5.0/parallel-netcdf-1.12.2-dnfq2vfdlwarq5tif4iqjm5va75644cd
      Setting Environment RS_PER_NODE=2
      Setting Environment SMT_MODE=1

    SUBMIT CMD:
      bsub -q batch -W 00:30 -P cli115 -env "all, ARGS_FOR_SCRIPT=--resubmit" .case.run

    MPIRUN (job=case.run):
      jsrun -X 1 --nrs 80 --rs_per_host 2 --tasks_per_rs 21 -d plane:21 --cpu_per_rs 21 --bind packed:smt:2 --latency_priority cpu-cpu --stdio_mode prepended   /gpfs/alpine/cli115/proj-shared/onguba/e3sm_scratch/v1-cpu-ne120-nnodes40-june06-check-master/bld/e3sm.exe   >> e3sm.log.$LID 2>&1 
```

**Comment by oksanaguba** [comment]
The script is essentially this
```
COMPSET=F2010-SCREAMv1      #or whatever compset you want
RES=ne120_ne120               #or whatever resolution you want
PECOUNT=12x1                # Number of MPIs by number of threads. Should be divisible by node size
CASE_NAME=v1-cpu-ne120-nnodes40-june06-check-master  #name for your simulation.
QUEUE=debug                 #whatever the name of your debug or batch queue is
WALLTIME=00:30              #HH:MM on Summit, HH:MM:SS on other machines
COMPILER=gnu               #which compiler to use (can be omitted on some machines)

~/${which}/cime/scripts/create_newcase -mach summit-cpu --case ${CASE_NAME} --compset ${COMPSET} --res ${RES} --compiler ${COMPILER} --walltime ${WALLTIME} --queue ${QUEUE}

cd ${CASE_NAME}

./xmlchange JOB_QUEUE=batch #debug if on cori or perlmutter
./xmlchange JOB_WALLCLOCK_TIME=0:30:00
./xmlchange STOP_OPTION=ndays #how long to run for
./xmlchange STOP_N=5

./xmlchange NTASKS=1680 #change how many MPI tasks to use
./xmlchange PIO_NETCDF_FORMAT="64bit_data"
```

**Comment by PeterCaldwell** [comment]
@sarats - can you comment on this?

**Comment by sarats** [comment]
Coincidentally, I brought up this topic (Summit settings) at our perf call yesterday. Some of the settings like the one above are hard-coded for the CPU cases by Az where SMT2 (2 hardware threads per core) is preferred. We plan to make an upstream issue/fix.

However, we use SMT1 in the GPU cases and my hunch is that this setting should not have much impact since it _should_ still bind to the same physical CPU core. The best way to confirm is to look at job step viewer output (waiting on the debug queue).

**Comment by sarats** [comment]
On an interactive node with SMT1, here are the placement figures. We need to check what SMT mode bsub is requesting as well for this run. 

For CPU cases with SMT1 mode, 
smt:1 gives what we wish to see for the for the CPU case in the case of non-threaded runs.

As I expected, the GPU cases look fine. 

CPU cases:

smt:1

![image](https://user-images.githubusercontent.com/5324834/172434451-2b4457bd-46a0-46d9-9185-216fd38cacc4.png)


smt:2 complains about 
`Warning: more than 1 task/rank assigned to a core` but I see a 'round-robin cyclic' placement than over-subscription of cores.


![image](https://user-images.githubusercontent.com/5324834/172434947-9f7d7c07-a83c-4985-b128-e6348da74991.png)


GPU cases:
![image](https://user-images.githubusercontent.com/5324834/172436079-83bfd712-0847-40f5-a180-6835bb7d1d9d.png)

![image](https://user-images.githubusercontent.com/5324834/172436187-49f3dc1c-85cd-4f5b-b8bb-616440f40753.png)

**Comment by sarats** [comment]
You can check in your .case.run file to see what SMT mode CIME is requesting.

**Comment by oksanaguba** [comment]
great, for now i hardcode "1" instead of "2". Not sure if there are plans to replace this logic with HT threads eventually, using OMP_NUM_THREADS variable below:
```
    <mpirun mpilib="spectrum-mpi">
      <executable>jsrun</executable>
      <arguments>
        <arg name="exit_on_error">-X 1</arg>
        <arg name="num_rs">$SHELL{if [ {{ total_tasks }} -eq 1 ];then echo --nrs 1 --rs_per_host 1;else echo --nrs $NUM_RS --rs_per_host $RS_PER_NODE;fi}</arg>
        <arg name="tasks_per_rs">--tasks_per_rs $SHELL{echo "({{ tasks_per_node }} + $RS_PER_NODE - 1)/$RS_PER_NODE"|bc}</arg>
        <arg name="distribute">-d plane:$SHELL{echo "({{ tasks_per_node }} + $RS_PER_NODE - 1)/$RS_PER_NODE"|bc}</arg>
        <arg name="cpu_per_rs">--cpu_per_rs $ENV{CPU_PER_RS}</arg>
        <arg name="task_bind">--bind packed:smt:1</arg>
        <arg name="latency_priority">--latency_priority $ENV{LTC_PRT}</arg>
        <arg name="stdio_mode">--stdio_mode prepended</arg>
        <arg name="thread_vars">$ENV{JSRUN_THREAD_VARS}</arg>
      </arguments>
```

**Comment by sarats** [comment]
Finally, as you are aware, I and others previously found E3SM workloads to get better performance with 84 MPI tasks per node on the CPU using the SMT2 mode. If desired, we need to figure out if that's the same for SCREAMv1. It's more of a % gain than 2x improvement.

On the other hand, I'm fine with running 42 MPIs/node for the INCITE benchmarks.

**Comment by amametjanov** [comment]
> Coincidentally, I brought up this topic (Summit settings) at our perf call yesterday. Some of the settings like the one above are hard-coded for the CPU cases by @amametjanov where SMT2 (2 hardware threads per core) is preferred. We plan to make an upstream issue/fix.

@sarats there is no hard-coding of SMT. If you look more carefully at `config_batch.xml`, `xmlchange MAX_TASKS_PER_NODE=42` can change default SMT2 to SMT1 (for CPU runs): refer to `.case.run` or `.case.test`.

**Comment by amametjanov** [comment]
> In config_machines there is line
 <arg name="task_bind">--bind packed:smt:2</arg>
Why is 2 hardcoded?

Just to clarify, this `jsrun` option has nothing to do with how many SMT slots per node are allocated in a job (`#BSUB -alloc_flags smtN` does that). This option is for placement/binding of OpenMP threads onto job-wide SMT slots.

In E3SM repo, `config_machines.xml` has it as `--bind packed:smt:$ENV{OMP_NUM_THREADS}`.
In scream repo, this was modified in a new `summit-cpu` machine-config in https://github.com/E3SM-Project/scream/commit/8472cac6f46e72a279da20e8bc8e9e0dc7555603 . Tagging @xyuan in case there is a specific reason why it's 2 in there.

**Comment by sarats** [comment]
Thanks @amametjanov for finding the cause.

Copy-pasting whole block in the recent PR meant that this line didn't show up as a difference. Moreover, the summit-gpu section retained the right setting for --bind. 
https://github.com/E3SM-Project/scream/pull/1687/files#diff-f8a781c6b7da95fb77cc79868123181ff46fc639c8fa9de5e60c1dfa8c15a659

**Comment by bartgol** [comment]
@sarats @oksanaguba I'm assuming this was fixed? Can I close?

**Comment by oksanaguba** [comment]
yes, can be closed -- summit is gone.

**Comment by PeterCaldwell** [comment]
Summit isn't entirely gone for a couple months though, right? I think Walter has some time there still... probably not relevant to SCREAM though, since we didn't get any more Summit time...

**Comment by bartgol** [comment]
Well, closing then. If folks run on Summit in the next couple of months and have issues, they'll likely open a new issue anyways...
